qmail-1.03/qscanq-stdin.8
.TH qscanq-stdin 8
.SH NAME
qscanq-stdin \- qmail virus scanner
.SH SYNOPSIS
.B qscanq-stdin
[
prog
arg1
arg2
arg3 ...
]

.SH DESCRIPTION
.B qscanq-stdin
is invoked by
.BR qscanq .
It ensures that the message on stdin is seekable, extracts its MIME components to the current
directory, and runs the virus scanner on them.

Like
.B qscanq
and
.BR qmail-multi ,
.B qscanq-stdin
needs no arguments. If arguments are supplied, they are used as the command to run instead of
.BR qmail-multi .
In this way, further qmail-multi wrappers--such as spam filters--can be chained with
.BR qscanq .

.B qscanq-stdin
expects an email message on file descriptor 1, its envelope on file descriptor 0, and writes
nothing to descriptor 2. It returns an exit code which is interpreted by the caller
(usually
.B qmail-smtpd
or
.B qmail-inject
).

The first thing
.B qscanq-stdin
does is to copy stdin to a file in the current directory named "msg". It then closes and reopens
the file read-only on file descriptor 0, and unlinks the file. This presents the same interface
to downstream programs while ensuring that the file is seekable, read-only, and not directly
accessible to other programs.

After making stdin seekable, 
.B qscanq-stdin
creates a second copy of the message in a file also named "msg". Some viruses are only detected by
looking at the raw message, so we need a copy in the working folder.

Once stdin is seekable,
.B qscanq-stdin
runs
.B RipMIME
to extract the MIME components of stdin into files in the current directory. 
.B RipMIME
is run with descriptors 1 and 2 closed, so it can't read the message envelope or print any output.
If
.B RipMIME
fails,
.B qscanq-stdin
returns temporary error code 71, resulting in message deferral. After
.B RipMIME
exits, stdin is rewound to the start of the file.

After
.B RipMIME
completes,
.B qscanq-stdin
runs
.B clamdscan
(or the scanner selected at compile time) to scan the current directory for viruses. All three
descriptors, 0, 1 and 2, are closed before running run-antivir. If it can't be run, or crashes,
then
.B qscanq-stdin
returns code 71, causing the message to be deferred. If
.B clamdscan
returns code 2, indicating that a virus has been found, then
.B clamdscan
returns error code 31, permanently rejecting the message.

If the virus scanner returns success, then
.B qscanq-stdin
will run the command supplied on its command line, or if none is supplied will run
.BR qmail-multi .
It also requests cleanup of its working directory by calling
.B run-cleanq

Upon startup,
.B qscanq-stdin
sets an alarm for the time specified in conf-ageout (default: 1 hour). When the alarm fires,
.B qscanq-stdin
sends SIGKILL to
.B RipMIME
or to the virus scanner, whichever may be running, and exits with code 71 (or code 54 if
.B ripmime
is killed). Once
.B qmail-queue
has started (invoked by
.B qmail-multi
), it sets its own alarm, independent of qscanq's alarm (default: 1 day).
.B qscanq
will wait forever if need be, so its longest runtime is equal to conf-ageout plus the timeout used
by
.BR qmail-queue .

.SH "CONTROL FILES"

.TP 5
.I badext
Unacceptable extensions for attachments.
.B qscanq-stdin
will reject mail if extensions of attachments in the mail matches lines in
.IR badext .
A line in
.I badext
may be of the form
.BR .\fIext ,
meaning any attachment having the extension
.IR ext .
The filename can be overriden by the environment variable BADEXT.
.TP 5
.I badextpatterns
Gives 
.B qscanq-stdin 
the ability to filter E-Mails by comparing the attachment extensions  
with a REGEX pattern in 
.IR badextpatterns . 
Example:

  *.doc
  !document.doc
  [0-9][0-9][0-9][0-9][0-9]@[0-9][0-9][0-9].doc
  *%*

.I badextpatterns
file with this contents stops all mail having attachment with extension .doc
except for document.doc. It also rejects all mail with extensions like:
12345@123.doc. Further, any E-Mail with a sender address containing a percent sign
(%) is rejected. The filename can be overriden by the environment variable BADEXTPATTERNS.

.SH Environment Variables

.B qscanq-stdin
sets no environment variables.
It does pass its environment to
.B qmail-queue
(or PROG, if supplied instead). This means that any qmail-queue wrapper must be able to handle its
environment safely.

If the DEBUG environment variable is set to any value whatsoever, then
.B qscanq-stdin
will print diagnostic messages to stderr whenever an error occurs. This is for use in debugging a
new configuration, and should never be used to capture "extra log information" during runtime.
If DEBUG is set for
.BR qscanq ,
then
.B qscanq
will set DEBUG=1 when calling
.BR qscanq-stdin .

It should be stressed that the value of DEBUG is ignored. Setting DEBUG to "1", "TRUE", "yes", etc.,
will turn on debug messages. Setting it to "0", "FALSE", "no", "N", etc., will also turn on debug
messages. If you don't want debug messages, don't set DEBUG at all.

The default scanner to be called is set during compilation in conf-scancmd. This can
be overriden at run time by the environment variable SCANCMD.

You can run
.B qscanq
with virus checkers other than
.BR clamdscan .
Just edit conf-scancmd with the
command-line you want
.B qscanq
to use. The command must obey the following rules:

1.  The current directory is recursively scanned.

2.  The scanner returns a 0 exit code if no virus is found.

3.  The scanner returns a 1 exit code if a virus is found.

4.  The scanner returns a 2 if some runtime error prevented the current directory from being scanned completely.

If your scanner's exit codes do not conform to the specification above, then you will need to
provide a wrapper which interprets your scanners return codes, and returns the above codes.
Then edit conf-scancmd or set SCANCMD environment variable with the command-line used to run your
wrapper.

.SH Return Codes

In addition to the return codes from
.BR qmail-queue ,
.B qscanq-stdin
returns 53 (write error) if it can't create a seekable copy of stdin, and 54 (read error) if it
can't reopen that file in read-only mode or if
.B RipMIME
fails. If
.B clamdscan
or
.B qmail-queue
fails, it returns code 71 (temporary rejection). If
.B clamdscan
returns exit code 2, then
.B qscanq-stdin
returns 31 (permanent rejection).

.SH "SEE ALSO"
qscanq(8),
cleanq(8),
run-cleanq(8),
qmail-queue(8),
qmail-multi(8),
clamd(8),
clamav.conf(5),
clamdscan(1),
clamscan(1)


qmail-1.03/qscanq-stdin.c
/*
 * $Log: qscanq-stdin.c,v $
 * Revision 1.4  2016-06-03 09:58:33+05:30  Cprogrammer
 * moved qmail-multi to sbin
 *
 * Revision 1.3  2005-02-14 23:06:19+05:30  Cprogrammer
 * unset VIRUSCHECK to prevent recursive calls to qscanq
 *
 * Revision 1.2  2004-10-22 20:29:47+05:30  Cprogrammer
 * added RCS id
 *
 * Revision 1.1  2004-09-22 23:26:46+05:30  Cprogrammer
 * Initial revision
 *
 */
#include <unistd.h>
#include <signal.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "auto_qmail.h"
#include "substdio.h"
#include "exit.h"
#include "open.h"
#include "fd.h"
#include "wait.h"
#include "seek.h"
#include "sig.h"
#include "auto_ageout.h"
#include "exitcodes.h"
#include "mkfn.h"
#include "auto_fnlen.h"
#include "env.h"
#include "strerr.h"

#define FATAL "qscanq: fatal: "

int             do_ripmime();
int             do_scan();
int             do_cleanq();

char            buf1[256];
substdio        ss1 = SUBSTDIO_FDBUF(write, 1, buf1, sizeof(buf1));
int             flaglog = 0;
int             alarm_flag;
pid_t           pid;

void
die_write()
{
	do_cleanq();
	_exit(53);
}

void
die_read()
{
	do_cleanq();
	_exit(54);
}

void
err(s)	/*- was named puts, but Solaris pwd.h includes stdio.h. dorks.  */
	char           *s;
{
	if (!flaglog)
		return;
	if (substdio_puts(&ss1, s) == -1)
		_exit(111);
}

/*- Signal handler just sets a flag.  */
void
alarm_handler(int sig)
{
	if (pid)
		kill(pid, SIGKILL);
	else
		_exit(QQ_XTEMP);
}

/*
 * The current directory is now "work", under a subdirectory of the
 * virus scanning queue. A message is on stdin, and its envelope on
 * file descriptor 2. Our environment is empty. Our job is to extract
 * the MIME parts of the message, and run the virus scanner in cwd.
 */
int
main(int argc, char *argv[])
{
	char           *(qqargs[]) = { "sbin/qmail-multi", 0 };
	int             fdout = -1;
	int             n = 0;
	char            inbuf[2048];
	struct substdio ssin;
	char            outbuf[256];
	struct substdio ssout;
	char            fn[FN_BYTES];
	pid = 0;

	/*- Set an alarm handler */
	alarm_flag = 0;
	sig_alarmcatch(alarm_handler);
	alarm(MAX_AGE);
	/*- Check whether we should be logging errors */
	if (env_get("DEBUG"))
		flaglog = 1;
	/* 
	 * Set up stdin with seekable copy of message
	 * [ fn := timestamp.ppid.n ]
	 */
	for(;;)
	{
		mkfn(fn, n++);
		/*- [ stdin := copied to a file named timestamp.ppid.n in cwd ] */
		if ((fdout = open(fn, O_RDWR|O_EXCL|O_CREAT, 0644)) < 0)
		{
			if (flaglog)
				strerr_die2sys(QQ_XTEMP, FATAL, "could not create file 'msg': ");
			do_cleanq();
			_exit(QQ_XWRTE);
		}
		if (n == 1)
			unlink(fn);
		substdio_fdbuf(&ssout, write, fdout, outbuf, sizeof(outbuf));
		substdio_fdbuf(&ssin, read, 0, inbuf, sizeof(inbuf));
		switch (substdio_copy(&ssout, &ssin))
		{
		case -2:
			die_read();
		case -3:
			die_write();
		}
		if (substdio_flush(&ssout) == -1)
			die_write();
		if (n == 2)
		{
			close(fdout);
			break;
		}
		if (fd_move(0, fdout) < 0)
		{
			if (flaglog)
				strerr_die2sys(QQ_XTEMP, FATAL, "fdmove failed: ");
			do_cleanq();
			_exit(QQ_XREAD);
		}
		seek_begin(0);
	}
	/*
	 * Run ripmime now, to extract the message parts.
	 * [ stdin := extracted to files under cwd ]
	 */
	seek_begin(0);
	if (!do_ripmime())
	{
		err("MIME extraction failed\n");
		do_cleanq();
		_exit(QQ_XREAD);
	}
	/*-
	 * Run antivir now, to check the current directory for viruses.
	 */
	seek_begin(0);
	switch (do_scan())
	{
	case (EX_ALLOK):
		break;
	case (EX_BADEX): /*- Found a banned extension */
		do_cleanq();
		_exit(QQ_BADEX); /*- drop the message */
	case (EX_VIRUS): /*- Found a virus */
		do_cleanq();
		_exit(QQ_VIRUS); /*- drop the message */
	default: /*- some problem w/scan */
		err("some problem occurred while scanning the message\n");
		do_cleanq();
		_exit(QQ_XTEMP);
	}
	/*- all done!  */
	do_cleanq();
	if (chdir(auto_qmail))
		_exit(61);
	/* 
	 * unset VIRUSCHECK to prevent
	 * qscanq getting called recursively
	 * by qmail-multi
	 */
	if (!env_unset("VIRUSCHECK"))
		_exit(51);
	if (argc > 1)
		execv(argv[1], argv + 1);
	else
		execv(*qqargs, argv);
	_exit(QQ_XTEMP);
}

void
getversion_qscanq_stdin_c()
{
	static char    *x = "$Id: qscanq-stdin.c,v 1.4 2016-06-03 09:58:33+05:30 Cprogrammer Exp mbhangui $";

	x++;
}
